### PR TITLE
fix: dedup RSC client references to prevent module duplication in dev

### LINKED ISSUE #409: Module duplication: client references resolve to raw ESM paths instead of pre-bundled paths
## Summary

When a React Server Component imports a `'use client'` module (e.g. from `@mantine/core`), vinext creates client references that resolve to the raw ESM path (`/node_modules/@mantine/core/esm/...`). Meanwhile, client components that directly import the same package load from Vite's pre-bundled path (`/node_modules/.vite/deps/@mantine_core.js`).

This creates **two separate module instances** in the browser, which means `createContext()` is called twice, producing different React context objects. Any context provider using the pre-bundled copy is invisible to consumers using the raw ESM copy, and vice versa.

## Reproduction

1. Set up a vinext project with `@mantine/core` (or any library that uses React context)
2. In the root layout, render a `MantineProvider` via a `'use client'` wrapper component
3. In a page (server component), import and render a Mantine component like `<TextInput />`
4. Open the page in the browser

**Expected:** TextInput renders correctly, finding MantineProvider in the component tree.

**Actual:** Error: `@mantine/core: MantineProvider was not found in component tree`

## Root cause

In Vite's dev mode, vinext runs three environments (RSC, SSR, client) with separate module graphs. When the RSC environment encounters a `'use client'` import, it serialises a client reference containing the module's raw file path. On the client, this reference is loaded as a raw ESM module from the Vite dev server.

However, the `MantineProvider` (rendered by a client component in the layout) loads `@mantine/core` through Vite's dependency pre-bundling, which bundles the package into a single `.vite/deps/@mantine_core.js` file.

The browser ends up loading both:
- `/node_modules/.vite/deps/@mantine_core.js` (pre-bundled, used by client components)
- `/node_modules/@mantine/core/esm/core/MantineProvider/MantineThemeProvider/MantineThemeProvider.mjs` (raw ESM, used by client references from server components)

These are separate JavaScript modules with separate `createContext()` calls, so the React context set by one is not visible to the other.

## Workaround

Exclude the affected packages from Vite's dependency pre-bundling so all imports use raw ESM paths consistently:

```ts
// vite.config.ts
const mantinePackages = [
  "@mantine/core",
  "@mantine/hooks",
  "@mantine/modals",
  "@mantine/notifications",
  "@mantine/dates",
  "@mantine/charts",
  "@mantine/nprogress",
  "@mantine/form",
];

// CJS-only transitive deps that browsers can't load as raw ESM
const mantineCjsDeps = [
  "prop-types",
  "loose-envify",
  "object-assign",
  "react-is",
  "js-tokens",
  "fast-deep-equal",
  "klona",
];

function environmentOptimizeDepsFix() {
  return {
    name: "environment-optimize-deps-fix",
    config(config) {
      if (!config.environments) return;
      for (const [envName, env] of Object.entries(config.environments)) {
        if (env?.optimizeDeps) {
          env.optimizeDeps.exclude = [
            ...(env.optimizeDeps.exclude || []),
            ...mantinePackages,
          ];
          // CJS deps must stay pre-bundled in the client (browser) environment
          if (envName === "client") {
            env.optimizeDeps.include = [
              ...(env.optimizeDeps.include || []),
              ...mantineCjsDeps,
            ];
          }
        }
      }
    },
  };
}
```

This forces all `@mantine` imports to use raw ESM, eliminating the duplication. The downside is significantly more HTTP requests in dev mode (500+ individual module files instead of one pre-bundled file).

## Suggested fix

Client references created by the RSC environment should resolve to the client environment's pre-bundled module paths (when available), rather than raw ESM file paths. This would ensure that both direct client imports and RSC client references load from the same module instance, preserving React context identity.

## Environment

- Vite: 7.3.1
- [@mantine/core](https://github.com/mantinedev/mantine): 8.x
- React: 19.x

### PR DESCRIPTION
**Pull request by Divkix**
## Summary

Fixes #409

When the RSC plugin creates `client-in-server-package-proxy` modules for `"use client"` submodules within packages (e.g., `@mantine/core`), it re-exports from absolute file paths. These bypass Vite's pre-bundling in the client environment, causing the browser to load two separate module instances — one from the raw ESM path and one from `.vite/deps/`. This breaks React context providers/consumers since `createContext()` runs twice.

- Add `vinext:client-reference-dedup` plugin (`enforce: "pre"`, dev-only) that intercepts absolute `node_modules` path imports originating from proxy virtual modules in the client environment and redirects them through bare specifier imports, which Vite's import analysis routes through pre-bundling automatically
- Respects `optimizeDeps.exclude` from resolved config so user overrides are honored
- No RSC plugin modifications needed — works transparently at the Vite plugin level

## Test plan

- [x] Unit tests for `extractPackageName` (regular, scoped, nested node_modules, edge cases)
- [x] Unit tests for plugin `resolveId`/`load` hooks (client env gating, proxy importer gating, exclude list)
- [x] Integration test with `fake-context-lib` fixture package (internal `"use client"` submodule with `createContext` — verifies provider value is readable by consumer)
- [x] `pnpm run typecheck` passes
- [x] `pnpm run lint` passes
- [x] `pnpm run fmt:check` passes

### CODE REVIEW

File: tests/client-reference-dedup.test.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+    it("skips non-absolute paths", () => {
+      const ctx = createContext("client");
+      const result = resolveId.call(
+        ctx,
+        "@mantine/core",
+        "\0virtual:vite-rsc/client-in-server-package-proxy/abc123",
+      );
+      expect(result).toBeUndefined();
+    });
```
[the remark this answers is not in the document]
Comment (Divkix) [reply tests/client-reference-dedup.test.ts:93]: Vite normalizes all paths to forward slashes internally — `resolveId` receives Vite-resolved IDs which are already POSIX-style regardless of host OS. The RSC plugin uses these normalized IDs too. The project doesn't target or CI on Windows currently. If Windows support is ever needed, it's a trivial follow-up to add `normalizePath()` at the entry point of `resolveId`.

Dismissing — not actionable for this PR.

File: packages/vinext/src/plugins/client-reference-dedup.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      handler(id, importer) {
+        // Only operate in the client environment
+        if (this.environment?.name !== "client") return;
+
+        // Only intercept imports from client-in-server-package-proxy modules
+        if (!importer || !importer.includes(PROXY_MARKER)) return;
+
+        // Only handle absolute paths through node_modules
+        if (!id.startsWith("/") || !id.includes("/node_modules/")) return;
+
```
[the remark this answers is not in the document]
Comment (Divkix) [reply packages/vinext/src/plugins/client-reference-dedup.ts:67]: Same as the other comment — Vite normalizes all module IDs to forward slashes before they reach plugin hooks. The `resolveId` hook never sees backslash-separated or drive-letter paths because Vite's internal resolution already normalizes them. The project doesn't target or CI on Windows.

Dismissing — not actionable for this PR.

File: tests/fixtures/app-basic/app/context-dedup-test/page.tsx [hunk 3]
```diff
@@ -0,0 +1,13 @@
+"use client";
+
+import { useTheme } from "fake-context-lib";
+
+export default function ContextDedupPage() {
+  const theme = useTheme();
+  return (
+    <div>
+      <h1>Context Dedup Test</h1>
+      <p data-testid="theme-value">Theme: {theme ?? "NOT_FOUND"}</p>
```
[the remark this answers is not in the document]
Comment (Divkix) [reply tests/fixtures/app-basic/app/context-dedup-test/page.tsx:10]: Valid observation — the test fixture wasn't actually exercising the RSC proxy path.

The fix is simpler than suggested: remove `"use client"` from the **layout**, not the page. The layout imports `ThemeProvider` from `fake-context-lib` — when the layout is a server component, RSC encounters `ThemeProvider`, follows the re-export chain to `internal/context.js` (which has `"use client"`), and creates a `client-in-server-package-proxy` with an absolute path. That's the exact code path the dedup plugin fixes.

The page stays `"use client"` — it imports `useTheme` in the client environment, which goes through pre-bundling with a bare specifier. This creates the duplication scenario: proxy absolute path (from RSC rendering the layout) vs pre-bundled bare specifier (from client rendering the page).

The suggested `ThemeConsumerClient` approach would put both sides through the proxy path, which doesn't reproduce the duplication bug.

Fixed by removing `"use client"` from `layout.tsx`.

File: tests/app-router.test.ts [hunk 4]
```diff
@@ -133,6 +133,21 @@ describe("App Router integration", () => {
...
+  // causing React context providers to be duplicated (createContext runs twice).
+  it("renders context provider/consumer from package with internal 'use client' submodule", async () => {
+    const { res, html } = await fetchHtml(baseUrl, "/context-dedup-test");
+    expect(res.status).toBe(200);
+    expect(html).toContain("Context Dedup Test");
+    // If module dedup is working, the consumer reads the provider's value.
+    // If broken, useContext returns null and we see "NOT_FOUND".
+    expect(html).toContain("dark-test-theme");
+    expect(html).not.toContain("NOT_FOUND");
+  });
```
[the remark this answers is not in the document]
Comment (Divkix) [reply tests/app-router.test.ts:149]: Valid — the integration test could pass even with the dedup plugin broken because both provider and consumer were in the client graph.

Fixed by removing `"use client"` from `layout.tsx`. Now the layout is a server component, so `ThemeProvider` goes through RSC → `client-in-server-package-proxy` (absolute path), while the page's `useTheme` goes through client pre-bundling (bare specifier). The test now genuinely validates that the dedup plugin resolves both to the same module instance.
